warehouse/query-core/src/main/java/datawave/query/planner/FederatedQueryPlanner.java

```diff
@@ -0,0 +1,315 @@
+import datawave.query.model.FieldIndexHole;
+import datawave.query.planner.pushdown.rules.PushDownRule;
+import datawave.query.tables.ScannerFactory;
+import datawave.query.util.MetadataHelper;
+import datawave.util.time.TraceStopwatch;
+import datawave.webservice.common.logging.ThreadConfigurableLogger;
+import datawave.webservice.query.Query;
+import datawave.webservice.query.configuration.GenericQueryConfiguration;
+import datawave.webservice.query.configuration.QueryData;
+
+public class FederatedQueryPlanner extends QueryPlanner {
+
+    private static final Logger log = ThreadConfigurableLogger.getLogger(FederatedQueryPlanner.class);
+    private final SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy/MM/dd");
+    private final Calendar calendar = Calendar.getInstance();
+
+    private final ShardQueryConfiguration originalConfig;
+    private final DefaultQueryPlanner originalPlanner;
+    private final Date originalBeginDate;
+    private final Date originalEndDate;
+    private String plannedScript;
+
+    public FederatedQueryPlanner(ShardQueryConfiguration config, DefaultQueryPlanner planner) {
+        this.originalConfig = config;
+        this.originalPlanner = planner;
+        this.originalBeginDate = config.getBeginDate();
+        this.originalEndDate = config.getEndDate();
+        Preconditions.checkNotNull(originalBeginDate, "Configuration begin date must not be null");
+        Preconditions.checkNotNull(originalEndDate, "Configuration end date must not be null");
+    }
+
+    @Override
+    public CloseableIterable<QueryData> process(GenericQueryConfiguration genericConfig, String query, Query settings, ScannerFactory scannerFactory)
+                    throws DatawaveQueryException {
+        // Validate the config type.
+        if (!(genericConfig instanceof ShardQueryConfiguration)) {
+            throw new ClassCastException("Config must be an instance of " + ShardQueryConfiguration.class.getSimpleName());
+        }
+
+        ShardQueryConfiguration config = (ShardQueryConfiguration) genericConfig;
+
+        log.debug("Query originally set to execute against date range " + dateFormat.format(originalBeginDate) + "-" + dateFormat.format(originalEndDate));
+
+        // Get the relevant date ranges.
+        // TODO - Determine if we should pass in fields and datatypes to filter on for this query. Can we do this before calling process() on the query?
+        SortedSet<Pair<Date,Date>> dateRanges = getValidTargetDates(Collections.emptySet(), Collections.emptySet());
+
+        // TODO - Determine how restrictive we should be when evaluating whether or not to retain a date in the target date range, i.e. should we refrain from
+        // querying on a date if any index holes are seen on that day (current implementation) or only when we see index holes on a date for all fields and
+        // datatypes established for a query? What about when a query will include all fields and/or all datatypes?
```

Comment:
I think we should only federate if the we see index holes for the fields in the query itself.  There is a visitor to return all the fields in the query, then apply the query model, and then you can determine the index holes to take into account.

Reply:
What's the name of the visitor class? Would we need a different approach for any fields specified via a regex pattern?

Reply:
What about an unfielded term that expands into a field with an index hole? It seems like we should take that case into account.

Reply:
oh right.... we would need to handle unfielded expansion.  hmm